relations/src/r1cs/constraint_system.rs

```diff
@@ -309,91 +310,111 @@ impl<F: Field> ConstraintSystem<F> {
             return;
         }
 
-        let used_times = self.lc_num_times_used(true);
-
-        // step 2: Modify the lcs accordingly
-        let mut outlined_lcs = BTreeMap::new();
-
-        let mut additional_a_constraints = Vec::<LcIndex>::new();
-        let mut additional_c_constraints_witness_assignments = Vec::<usize>::new();
+        let mut num_times_used = self.lc_num_times_used(false);
 
+        let mut new_witness_linear_combinations = Vec::new();
+        let mut new_witness_indices = Vec::new();
+
+        // `new_lc_msp` stores the modified linear combinations.
+        //
+        // This loop goes through all the LCs in the map, starting from
+        // the early ones, and decides whether or not to dedicate a witness
+        // variable for this LC.
+        //
+        // If true, the LC is replaced with 1 * this witness variable.
+        // Otherwise, the LC is inlined.
+        //
+        let mut new_lc_msp = BTreeMap::new();
         for (&index, lc) in self.lc_map.iter() {
-            let mut outlined_lc = LinearCombination::new();
+            // Inline the LC.
+            let mut inlined_lc = LinearCombination::new();
             for &(coeff, var) in lc.iter() {
                 if var.is_lc() {
                     let lc_index = var.get_lc_index().expect("should be lc");
+
                     // If `var` is a `SymbolicLc`, fetch the corresponding
                     // inlined LC, and substitute it in.
-                    let lc = outlined_lcs.get(&lc_index).expect("should be inlined");
-                    outlined_lc.extend((lc * coeff).0.into_iter());
+                    //
+                    // We have the guarantee that `lc_index` must exist in
+                    // `new_lc_msp` since a LC can only depend on other
```

Comment:
msp -> map? (Here and elsewhere?)

Reply:
Ah! Working on it.

I need to be careful. Recently my typo rate is very high.